Infer compression from non-string paths

dhimmel opened the pull request:
Currently, `compression='infer'` only works on paths that are strings. However, pandas supports non-string paths, such as `pathlib.Path` objects. This PR adds support for inferring compression for all path types (for io interfaces that use `io.common._infer_compression`).

## Checklist

- [x] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
- [x] No issue opened
- [x] tests added / passed
- [x] whatsnew entry

dhimmel committed:
Test: infer compression from pathlib.Path

gfyoung commented on pandas/tests/io/parser/compression.py:
```diff
@@ -166,6 +166,21 @@ def test_read_csv_infer_compression(self):
+        """
+        Test that compression is inferred from pathlib.Path paths.
+        """
+        try:
```
Does this test pass on `master` (or are you going to make changes so that it will) ?

dhimmel replied:
As of https://github.com/pandas-dev/pandas/pull/17206/commits/a3b9ee9b558588bf2973aa90c5d12602799c1bb1, this test should fail when `pathlib` is available and be skipped otherwise.

Once I commit the enhancement, this test should pass when  `pathlib` is available and be skipped otherwise.

TomAugspurger replied:
What's your plan for the fix? We do have the `fspath` stuff now. It's possible we could just ensure that the path has been stringified before doing the compression inference.

dhimmel replied:
@TomAugspurger see https://github.com/pandas-dev/pandas/pull/17206/commits/b4e280f451cfd1f02954fd02667a7bc9d3271263. The fix should be simple I think, just passing `filepath_or_buffer` through the `_stringify_path` function in `_infer_compression`.

TomAugspurger replied:
Yep, perfect (the fspath stuff happens inside `_stringify_path`). Let's hope that works.

gfyoung replied:
Separately, let's take advantage of `pytest.importorskip` functionality here instead of doing this `try-except` block.

dhimmel committed:
Infer compression from non-string paths

dhimmel commented:
@TomAugspurger @gfyoung, we're passing. Did you want me to make a What's New entry?

gfyoung commented:
Yes, absolutely!

dhimmel committed:
Documentation updates

gfyoung commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -126,6 +126,7 @@ Other Enhancements
 - :func:`date_range` now accepts 'Y' in addition to 'A' as an alias for end of year (:issue:`9313`)
 - Integration with `Apache Parquet <https://parquet.apache.org/>`__, including a new top-level :func:`read_parquet` and :func:`DataFrame.to_parquet` method, see :ref:`here <io.parquet>`.
 - :func:`DataFrame.add_prefix` and :func:`DataFrame.add_suffix` now accept strings containing the '%' character. (:issue:`17151`)
+- :func:`read_csv` can now infer compression from non-string paths, such as a ``pathlab.Path`` objects (:issue:`17206`).
```
I think you meant `pathlib` instead  of `pathlab`

jreback replied:
say ``read_*`` methods (as this is a general fix)

gfyoung commented on pandas/tests/io/parser/compression.py:
```diff
@@ -166,6 +166,21 @@ def test_read_csv_infer_compression(self):
+        except ImportError:
+            pytest.skip('need pathlib to run')
+        expected = self.read_csv(self.csv1, index_col=0, parse_dates=True)
+        for extension in '', '.gz', '.bz2':
```
Let's use `pytest.mark.parametrize` decorator for the extension.

gfyoung commented on pandas/tests/io/parser/compression.py:
```diff
@@ -166,6 +166,21 @@ def test_read_csv_infer_compression(self):
+            import pathlib
+        except ImportError:
+            pytest.skip('need pathlib to run')
+        expected = self.read_csv(self.csv1, index_col=0, parse_dates=True)
```
Construct the expected `DataFrame` from the `DataFrame` constructor, not from `read_csv`.

dhimmel replied:
I'm not sure how to do that... I followed the method used by other tests in this module, e.g. `test_read_csv_infer_compression`. Is `csv1` available in memory somewhere? I agree that it would be nice to reduce the runtime burden of this test / dependencies on a file... But, it's not something I should manually create, I don't think.

jreback commented on pandas/tests/io/parser/compression.py:
```diff
@@ -166,6 +166,21 @@ def test_read_csv_infer_compression(self):
+        for extension in '', '.gz', '.bz2':
+            path = pathlib.Path(self.csv1 + extension)
+            df = self.read_csv(
+                path, index_col=0, parse_dates=True, compression='infer')
```
I would remove this test entirely. instead add a generalized compression test (iow for all extensions & path types) in ``test_common``, similar to how fspath is tested (you specificy all parameters then you may skip on some combinations).

dhimmel replied:
@jreback, Just to make sure we're on the same page. All path types would be:

+ local paths via strings
+ local paths via __fspath__ defining objects
+ URLs

All compression types would be None, xz, bz2, zip, and gzip. But in all cases, we'd be inferring compression?

Just want to make sure the scope of this test should be primarily `compression='infer'`? Any more guidance will be appreciated.

dhimmel committed:
Address @gfyoung reviews

https://github.com/pandas-dev/pandas/pull/17206#pullrequestreview-55411442
https://github.com/pandas-dev/pandas/pull/17206#pullrequestreview-55411536
https://github.com/pandas-dev/pandas/pull/17206#discussion_r132341564

dhimmel committed:
Generalize What's New entry

https://github.com/pandas-dev/pandas/pull/17206#pullrequestreview-55495655

dhimmel commented:
Currently, this PR is ready, except for @jreback's outstanding review: https://github.com/pandas-dev/pandas/pull/17206#pullrequestreview-55496287. I'm happy to "add a generalized compression test", but I need some additional guidance. Refer to the line-by-line discussion under the review.

TomAugspurger commented:
Moving that out of the inline comment that was squashed:

> @jreback, Just to make sure we're on the same page. All path types would be:
> 
> - local paths via strings
> - local paths via fspath defining objects
> - URLs
> All compression types would be None, xz, bz2, zip, and gzip. But in all cases, we'd be inferring > compression?

@dhimmel it'll be similar to https://github.com/pandas-dev/pandas/blob/330b8c1c195174f729a1d2ee6f916ebd1579217e/pandas/tests/io/test_common.py#L130

You can parametrize over readers, extensions, and path types (string, pathlib, py.path, and `CustomFSPath`).

What all readers should this support?

jreback commented:
quite a lot of these tests already exists

just the case for the oathlib type objects i think are a bit lite

however the fspath tests are centralized 
i would move these to the io readers themselves (where the reg path compression tests are)

dhimmel committed:
Test _infer_compression in io/test_common.py

dhimmel commented:
Okay I moved the tests to `tests/io/test_common.py`. The new test only tests `_infer_compression` rather than testing entire file reading, so it's lighter weight. 

https://github.com/pandas-dev/pandas/blob/0f925c15d6d8a3c2248139f07fef3bb61385fecd/pandas/tests/io/test_common.py#L91-L102

> You can parametrize over ... path types (string, pathlib, py.path, and CustomFSPath).

@TomAugspurger see what I did in https://github.com/pandas-dev/pandas/pull/17206/commits/0f925c15d6d8a3c2248139f07fef3bb61385fecd to parametrize over path types. It requires creating a module level variable `path_types`... I couldn't identify a more elegant way.

dhimmel committed:
fixup! Generalize What's New entry

TomAugspurger reviewed (approved):
@dhimmel thanks, that looks good. I agree that the module-level `path_types` isn't the most elegant. You could *potentially* add that to `conftest.py`, but I think it's fine as is.

I also prefer that the tests only hit `_infer_compression`, as long as we have at least some tests actually doing the full IO (which I believe we do).

dhimmel commented:
> You could potentially add that to conftest.py, but I think it's fine as is.

Probably best for me to avoid touching `pandas/conftest.py`. Let's save that for someone who knows the the testing architecture a bit better (:

gfyoung commented:
> You could potentially add that to conftest.py, but I think it's fine as is.

I agree.  There is no need to do this.

gfyoung commented:
Thanks @dhimmel !

jreback commented:
@gfyoung pls don't merge things that I have an outstanding review on.

jreback commented:
> - `read_*` methods can now infer compression from non-string paths, such as ``pathlib.Path`` objects (:issue:`17206`).

If this is claimed, then this needs tests for each type (e.g. html, pytables, prob some others).

@dhimmel pls post a new PR with this.

@gfyoung pls create a new issue to track this fix.

dhimmel commented:
> pls post a new PR with this

Okay the actual scope of this PR is `read` methods that infer compression using `pandas.io.common._infer_compression`. Would that be appropriate to state in the What's New? 

Ideally, all read methods would use the common functions in `io.common`, but that's not the case (see https://github.com/pandas-dev/pandas/issues/15008). I don't really have time right now to go through all these read methods and see which are affected. I am interesting in consolidating the `read_*` code, but would prefer to wait till `pandas2`... since ensuring Python 2 support makes this work quite tedious.

Given these considerations, this PR's scope is to add path-like object support to `_infer_compression`. @jreback let me know how to update the What's New.
